packages/ssz/src/type/stableContainer.ts

```diff
@@ -0,0 +1,794 @@
+      if (offset !== fixedEnd) {
+        throw new Error(`First offset must equal to fixedEnd ${offset} != ${fixedEnd}`);
+      }
+    } else {
+      if (offset < offsets[i - 1]) {
+        throw new Error(`Offsets must be increasing ${offset} < ${offsets[i - 1]}`);
+      }
+    }
+
+    offsets[i] = offset;
+  }
+
+  return offsets;
+}
+
+/**
+ * Precompute sizes of the Container doing one pass over fields
+ */
+function precomputeSizes<Fields extends Record<string, Type<unknown>>>(
+  fields: FieldEntry<Fields>[]
+): {
+  minLen: number;
+  maxLen: number;
+  fixedSize: number | null;
+} {
+  // at a minimum, the active fields bitvector is prepended
+  const activeFieldsLen = Math.ceil(fields.length / 8);
+
+  let minLen = activeFieldsLen;
+  let maxLen = activeFieldsLen;
+  const fixedSize = null;
+
+  for (const {fieldType, optional} of fields) {
+    minLen += optional ? 0 : fieldType.minSize;
+    maxLen += fieldType.maxSize;
+
+    if (fieldType.fixedSize === null) {
+      // +4 for the offset
+      minLen += optional ? 0 : 4;
+      maxLen += 4;
+    }
+  }
+  return {minLen, maxLen, fixedSize};
+}
+
+/**
+ * Compute the JSON key for each fieldName. There will exist a single JSON representation for each type.
+ * To transform JSON payloads to a casing that is different from the type's defined use external tooling.
+ */
+export function precomputeJsonKey<Fields extends Record<string, Type<unknown>>>(
+  fieldName: keyof Fields,
+  casingMap?: CasingMap<Fields>,
+  jsonCase?: KeyCase
+): string {
+  if (casingMap) {
+    const keyFromCaseMap = casingMap[fieldName];
+    if (keyFromCaseMap === undefined) {
+      throw Error(`casingMap[${fieldName}] not defined`);
+    }
+    return keyFromCaseMap as string;
+  } else if (jsonCase) {
+    return Case[jsonCase](fieldName as string);
+  } else {
+    return fieldName as string;
+  }
+}
+
+/**
+ * Render field typeNames for a detailed typeName of this Container
+ */
+export function renderContainerTypeName<Fields extends Record<string, Type<unknown>>>(
+  fields: Fields,
+  prefix = "StableContainer"
+): string {
+  const fieldNames = Object.keys(fields) as (keyof Fields)[];
+  const fieldTypeNames = fieldNames.map((fieldName) => `${fieldName}: ${fields[fieldName].typeName}`).join(", ");
+  return `${prefix}({${fieldTypeNames}})`;
+}
+
+/**
+ * Get the active field bitvector, given the root of the tree and # of fields
+ */
+export function getActiveFields(rootNode: Node, bitLen: number): BitArray {
+  // fast path for depth 1, the bitvector fits in one chunk
+  if (bitLen <= 256) {
+    return new BitArray(rootNode.right.root.subarray(0, Math.ceil(bitLen / 8)), bitLen);
+  }
+
+  const activeFieldsBuf = new Uint8Array(Math.ceil(bitLen / 8));
+  const depth = countToDepth(BigInt(activeFieldsBuf.length));
+  const nodes = getNodesAtDepth(rootNode.right, depth, 0, Math.ceil(bitLen / 256));
+  for (let i = 0; i < nodes.length; i++) {
+    activeFieldsBuf.set(nodes[i].root, i * 32);
```

Comment:
I don't understand this, `activeFieldsBuf` is allocated as `bitLen / 8` so it's not supposed to store root of a node. Given there are 8 fields, bitLen = 8 so the `activeFieldsBuf` is exactly 1 byte

I suppose each bit is respective to 1 field, so if node is `Bytes32()` it's a `true`, otherwise it's a `false`?

Reply:
this is supposed to not to get to the data nodes, but the activeFileds nodes

```
const depth = countToDepth(BigInt(activeFieldsBuf.length));const depth = countToDepth(BigInt(activeFieldsBuf.length));
```